### PR TITLE
Add material properties into Richards Flow Process

### PR DESCRIPTION
**Pull request by Yonghui56**
As title.
This work is still work in progress.
Document stuff is still missing.

### CODE REVIEW

**Comment by Yonghui56** [comment]
Documents have been included.
Next step the ctest will be updated.

**Comment by Yonghui56** [comment]
Please review.

File: ProcessLib/RichardsFlow/RichardsFlowMaterialProperties.cpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+#include "MeshLib/Mesh.h"
+#include "MeshLib/PropertyVector.h"
+#include "ProcessLib/Parameter/Parameter.h"
+#include "ProcessLib/Parameter/SpatialPosition.h"
+
+namespace ProcessLib
+{
+namespace RichardsFlow
+{
+    RichardsFlowMaterialProperties::RichardsFlowMaterialProperties(
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowMaterialProperties.cpp]: @Yonghui56 @wenqing I have seen this type of class in every multi-phase flow process implementations and for the most part this is repeating the same things again and again. Can this redundancy be removed?

File: ProcessLib/RichardsFlow/RichardsFlowMaterialProperties.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+{
+    /** This class has a collection of material properties for Richards flow model
+    *  and it makes description of the properties of unsaturated porous media
+    *  i.e. the fluid density and viscosity models
+    *  the relative permeability models,
+    *  the capillary pressure-saturation relationships.
+    *  It generally provides the computation of the PDE coefficients for Richards flow.
+    */
+
+class RichardsFlowMaterialProperties
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowMaterialProperties.h]: final

File: ProcessLib/RichardsFlow/RichardsFlowProcess.cpp [hunk 3]
```diff
@@ -25,7 +25,11 @@ RichardsFlowProcess::RichardsFlowProcess(
     std::vector<std::reference_wrapper<ProcessVariable>>&& process_variables,
     RichardsFlowProcessData&& process_data,
     SecondaryVariableCollection&& secondary_variables,
-    NumLib::NamedFunctionCaller&& named_function_caller)
+    NumLib::NamedFunctionCaller&& named_function_caller,
+    BaseLib::ConfigTree const& /*config*/,
+    std::map<std::string,
+    std::unique_ptr<MathLib::PiecewiseLinearInterpolation>> const&
+    /*curves*/)
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowProcess.cpp]: Is this clang-format's formatting?

File: ProcessLib/RichardsFlow/RichardsFlowProcessData.h [hunk 4]
```diff
@@ -8,7 +8,7 @@
  */
 
 #pragma once
-
+#include "RichardsFlowMaterialProperties.h"
 namespace MeshLib
 {
 class Element;
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowProcessData.h]: Why is this forward-declaration needed?

**Comment by endJunction** [comment]
The commit messages could be more explanatory. That a commit is an update to something is already expressed by the commit's existence.

**Comment by endJunction** [comment]
There are three messages in doxygen on the Input Parameter -- QA page regarding Richards flow.

File: Documentation/ProjectFile/prj/processes/process/RICHARDS_FLOW/material_property/porous_medium/porous_medium/a_id.md [hunk 5]
```diff
@@ -0,0 +1,2 @@
+It specifies the material ID of a porous medium. The material IDs of elements
+should be given in the mesh's cell data array "MaterialIDs".
```
Comment (TomFischer) [inline Documentation/ProjectFile/prj/processes/process/RICHARDS_FLOW/material_property/porous_medium/porous_medium/a_id.md]: If the data array 'MaterialIDs' does not exist in the mesh then ogs crashs. For this reason, i would write instead of 'should' either 'must' or 'have'.

File: Documentation/ProjectFile/prj/processes/process/RICHARDS_FLOW/t_mass_lumping.md [hunk 6]
```diff
@@ -1 +1 @@
-\ogs_missing_documentation
+A tag for using mass-lumping scheme
```
Comment (TomFischer) [inline Documentation/ProjectFile/prj/processes/process/RICHARDS_FLOW/t_mass_lumping.md]: I'm not familiar with mass-lumping scheme. Is this tag for enabling/disabling mass-lumping scheme or is it used for choosing a specific mass-lumping scheme? It would be nice if the documentation is more meaningful.

File: Documentation/ProjectFile/prj/processes/process/RICHARDS_FLOW/material_property/porous_medium/porous_medium/a_id.md (hunk 5 above)
Comment (Yonghui56) [inline Documentation/ProjectFile/prj/processes/process/RICHARDS_FLOW/material_property/porous_medium/porous_medium/a_id.md]: Yes, without data array 'MaterialIDs'  would not lead to crash, it will return 0 for all material_id instead. The document will be updated.

Comment (endJunction) [inline Documentation/ProjectFile/prj/processes/process/RICHARDS_FLOW/material_property/porous_medium/porous_medium/a_id.md]: Could you please check if this change is also required in other processes?

Comment (TomFischer) [inline Documentation/ProjectFile/prj/processes/process/RICHARDS_FLOW/material_property/porous_medium/porous_medium/a_id.md]: The LiquidFlow, HT and ComponentTransport processes use a 'homogeneous' mode if the PropertyVector 'MaterialIDs' is not available in the mesh properties. In my opinion ogs should crash in this case. But this is/should not be part of this PR.

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 7]
```diff
@@ -92,75 +138,80 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
+                _process_data.material->getSaturationDerivative(
+                    material_id, t, pos, p_int_pt, temperature, Sw);
+
             // \TODO Extend to pressure dependent density.
             double const drhow_dp(0.0);
 
             double const mass_mat_coeff =
-                storage * Sw + poro * Sw * drhow_dp - poro * dSwdPc;
-            double const K_mat_coeff = K * k_rel / mu;
+                porosity * Sw * drhow_dp - porosity * dSw_dpc;
```
Comment (TomFischer) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: Compared to the previous version the term `storage * Sw` is missing. Is the new implementation correct?

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 8]
```diff
@@ -92,75 +138,80 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
-            local_K.noalias() += sm.dNdx.transpose() * K_mat_coeff * sm.dNdx *
-                                 sm.detJ * sm.integralMeasure * wp.getWeight();
+            local_M.noalias() += mass_mat_coeff * _ip_data[ip].mass_operator;
+
+            double const k_rel =
+                _process_data.material->getRelativePermeability(
+                    t, pos, p_int_pt, temperature, Sw);
+            auto const mu = _process_data.material->getFluidViscosity(
+                p_int_pt, temperature);
+            laplace_operator.noalias() = _ip_data[ip].dNdx.transpose() *
```
Comment (TomFischer) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: Maybe it is possible to update the matrix `local_K` directly and omit the creation, filling with zeros and setting the values of `laplace_operator` here?

File: ProcessLib/RichardsFlow/RichardsFlowMaterialProperties.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+#include "BaseLib/reorderVector.h"
+#include "MaterialLib/Fluid/FluidProperty.h"
+#include "MaterialLib/PorousMedium/Porosity/Porosity.h"
+#include "MaterialLib/PorousMedium/Storage/Storage.h"
+#include "MaterialLib/PorousMedium/UnsaturatedProperty/CapillaryPressure/CapillaryPressureSaturation.h"
+#include "MaterialLib/PorousMedium/UnsaturatedProperty/CapillaryPressure/CreateCapillaryPressureModel.h"
+#include "MaterialLib/PorousMedium/UnsaturatedProperty/RelativePermeability/CreateRelativePermeabilityModel.h"
+#include "MaterialLib/PorousMedium/UnsaturatedProperty/RelativePermeability/RelativePermeability.h"
+#include "MathLib/InterpolationAlgorithms/PiecewiseLinearInterpolation.h"
```
Comment (TomFischer) [inline ProcessLib/RichardsFlow/RichardsFlowMaterialProperties.cpp]: I think you don't need this include here.

**Comment by endJunction** [comment]
@Yonghui56 The missing storage term was added; I would guess the expected results (and potentially the graphs in the documentation) need an update.
Why was this issue not detected by the ctests?

**Comment by Yonghui56** [comment]
@endJunction because in the test, the storage term is 0. (that is why I eliminated this term firstly...)

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 10]
```diff
@@ -108,59 +149,67 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
-            double const Sw = interpolated_Pc.getValue(Pc);
+            double p_int_pt = 0.0;
+            NumLib::shapeFunctionInterpolate(local_x, _ip_data[ip].N, p_int_pt);
+            const double& temperature = _process_data.temperature(t, pos)[0];
+            auto const porosity = _process_data.material->getPorosity(
+                material_id, t, pos, p_int_pt, temperature, 0);
+
+            double const pc_int_pt = -p_int_pt;
+
+            double const Sw = _process_data.material->getSaturation(
```
Comment (wenqing) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: Change it to:
```
            double const Sw = (pc_int_pt>0) ?  _process_data.material->getSaturation(
                material_id, t, pos, p_int_pt, temperature, pc_int_pt) : 1.0;
```

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 11]
```diff
@@ -108,59 +149,67 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
+                material_id, t, pos, p_int_pt, temperature, pc_int_pt);
             _saturation[ip] = Sw;
 
-            double const k_rel = interpolated_Kr.getValue(Sw);
+            double const dSw_dpc =
```
Comment (wenqing) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: Change it to
```
            double const dSw_dpc = (pc_int_pt>0) ? 
                _process_data.material->getSaturationDerivative(
                    material_id, t, pos, p_int_pt, temperature, Sw) : 0.;
```

File: ProcessLib/RichardsFlow/RichardsFlowMaterialProperties.h [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+/** This class has a collection of material properties for Richards flow model
+*  and it makes description of the properties of unsaturated porous media
+*  i.e. the fluid density and viscosity models
+*  the relative permeability models,
+*  the capillary pressure-saturation relationships.
+*  It generally provides the computation of the PDE coefficients for Richards
+* flow.
+*/
+
+class RichardsFlowMaterialProperties final
```
Comment (wenqing) [inline ProcessLib/RichardsFlow/RichardsFlowMaterialProperties.h]: Future work could be to create a class for single phase fluid flow properties, and it has common fluid property data for LiquidFlow, GroundWaterFlow and RichardsFlow.

**Comment by Yonghui56** [comment]
Please review.

**Comment by wenqing** [comment]
:fast_forward:

File: ProcessLib/RichardsFlow/CreateRichardsFlowProcess.cpp [hunk 13]
```diff
@@ -122,10 +55,49 @@ std::unique_ptr<Process> createRichardsFlowProcess(
...
+        createRichardsFlowMaterialProperties(mat_config, material_ids);
+    RichardsFlowProcessData process_data{std::move(material),
+                                         specific_body_force, has_gravity,
+                                         mass_lumping, temperature};
+
     return std::make_unique<RichardsFlowProcess>(
         mesh, std::move(jacobian_assembler), parameters, integration_order,
         std::move(process_variables), std::move(process_data),
-        std::move(secondary_variables), std::move(named_function_caller));
+        std::move(secondary_variables), std::move(named_function_caller),mat_config,curves );
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/CreateRichardsFlowProcess.cpp]: clang-format?

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 14]
```diff
@@ -108,59 +149,67 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
+            auto const storage = _process_data.material->getStorage(
+                    material_id, t, pos, p_int_pt,temperature,0);
             double const mass_mat_coeff =
-                storage * Sw + poro * Sw * drhow_dp - poro * dSwdPc;
-            double const K_mat_coeff = K * k_rel / mu;
+                storage * Sw + porosity * Sw * drhow_dp - porosity * dSw_dpc;
 
-            local_K.noalias() += sm.dNdx.transpose() * K_mat_coeff * sm.dNdx *
-                                 sm.detJ * sm.integralMeasure * wp.getWeight();
+            local_M.noalias() += mass_mat_coeff * _ip_data[ip].mass_operator;
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: Add a note, that for a anisotropic `mass_mat_coeff` the `mass_operator` be changed, and which way.

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 15]
```diff
@@ -108,59 +149,67 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
-            local_M.noalias() += sm.N.transpose() * mass_mat_coeff * sm.N *
-                                 sm.detJ * sm.integralMeasure * wp.getWeight();
             if (_process_data.has_gravity)
             {
-                auto const rho_w = _process_data.water_density(t, pos)[0];
-                auto const body_force =
-                    _process_data.specific_body_force(t, pos);
+                auto const rho_w = _process_data.material->getFluidDensity(
+                    p_int_pt, temperature);
+                auto const body_force = _process_data.specific_body_force;
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: reference is sufficient here, isn't it?

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 16]
```diff
@@ -177,13 +226,18 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
         pos.setElementID(_element.getID());
-
-        auto const K = _process_data.intrinsic_permeability(t, pos)[0];
-        MathLib::PiecewiseLinearInterpolation const& interpolated_Pc =
-            _process_data.interpolated_Pc;
-        MathLib::PiecewiseLinearInterpolation const& interpolated_Kr =
-            _process_data.interpolated_Kr;
-        auto const mu = _process_data.viscosity(t, pos)[0];
+        const int material_id =
+            _process_data.material->getMaterialID(pos.getElementID().get());
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: minor: boost::optional also has `operator*` for access. One also could use `_element.getID()` as it is written two lines above.

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h (hunk 14 above)
Comment (Yonghui56) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: how is the mass_operator changed?

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 17]
```diff
@@ -193,27 +247,28 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
-            double const Pc = -P_int_pt;
-            double const Sw = interpolated_Pc.getValue(Pc);
-            double const k_rel = interpolated_Kr.getValue(Sw);
-            double const K_mat_coeff = K * k_rel / mu;
-
-            GlobalDimVectorType velocity =
-                -K_mat_coeff * sm.dNdx * p_nodal_values;
+            double p_int_pt = 0.0;
+            NumLib::shapeFunctionInterpolate(local_x, _ip_data[ip].N, p_int_pt);
+            double const pc_int_pt = -p_int_pt;
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: Is the extra variable necessary, or can `-p_int_pt` be used below?

Comment (Yonghui56) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: I would prefer to sticking to this.

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h [hunk 18]
```diff
@@ -193,27 +247,28 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
...
+                    p_int_pt, temperature);
+                auto const body_force = _process_data.specific_body_force;
                 assert(body_force.size() == GlobalDim);
-                auto const b = MathLib::toVector<GlobalDimVectorType>(
-                    body_force, GlobalDim);
-                // here it is assumed that the vector b is directed 'downwards'
-                velocity += K_mat_coeff * rho_w * b;
+                // here it is assumed that the vector body_force is directed
+                // 'downwards'
+                velocity += permeability * (k_rel / mu) * rho_w * body_force;
```
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: For the performance concerned programmers:
 - permeability * k_rel/mu is always used together, so it can be evaluated once.
 - velocity is always multiplied with `permeability * k_rel / mu`, so may be it can be done once after the if-condition.

**Comment by endJunction** [comment]
There are unrelated changes to [Jenkinsfile](https://github.com/ufz/ogs-data/compare/a80cf646e689b49f7202a1a642d7bf72236b70c2...1b30c432bbe45271710b54e83f84ad4f5d75a033#diff-58231b16fdee45a03a4ee3cf94a9f2c3); you might need to rebase the Tests/Data branch.

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h (hunk 14 above)
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: Sorry for my bad English. I mean that one would need to write dNdx^T * mass_coeff * dNdx if mass_coeff becomes anisotropic matrix, isn't it?

File: ProcessLib/RichardsFlow/RichardsFlowFEM.h (hunk 17 above)
Comment (endJunction) [inline ProcessLib/RichardsFlow/RichardsFlowFEM.h]: OK.

**Comment by endJunction** [comment]
When Tests/Data is no longer touching unrelated files, :fast_forward:

**Comment by Yonghui56** [comment]
@endJunction, I am afraid ogs/master has not included the changes in ogs-data/master yet
please have a look:
https://github.com/ufz/ogs-data/tree/a80cf646e689b49f7202a1a642d7bf72236b70c2
and 
https://github.com/ufz/ogs-data
or do I miss something?

**Comment by endJunction** [comment]
Yes, we missed to update Tests/Data pointer; Fixed now.
